Mathlib/Algebra/Exact.lean

```diff
@@ -40,57 +40,152 @@ def Exact [Zero P] : Prop := ∀ y, g y = 0 ↔ y ∈ Set.range f
 
 variable {f g}
 
-lemma Exact.apply_apply_eq_zero [Zero P] (h : Exact f g) (x : M) :
+namespace Exact
+
+lemma apply_apply_eq_zero [Zero P] (h : Exact f g) (x : M) :
     g (f x) = 0 := (h _).mpr <| Set.mem_range_self _
 
-lemma Exact.comp_eq_zero [Zero P] (h : Exact f g) : g.comp f = 0 :=
+lemma comp_eq_zero [Zero P] (h : Exact f g) : g.comp f = 0 :=
   funext h.apply_apply_eq_zero
 
-lemma Exact.of_comp_of_mem_range [Zero P] (h1 : g ∘ f = 0)
+lemma of_comp_of_mem_range [Zero P] (h1 : g ∘ f = 0)
     (h2 : ∀ x, g x = 0 → x ∈ Set.range f) : Exact f g :=
   fun y => Iff.intro (h2 y) <|
     Exists.rec ((forall_apply_eq_imp_iff (p := (g · = 0))).mpr (congrFun h1) y)
 
-lemma Exact.comp_injective [Zero P] [Zero P'] (exact : Exact f g)
+lemma comp_injective [Zero P] [Zero P'] (exact : Exact f g)
     (inj : Function.Injective g') (h0 : g' 0 = 0) :
     Exact f (g' ∘ g) := by
   intro x
   refine ⟨fun H => exact x |>.mp <| inj <| h0 ▸ H, ?_⟩
   intro H
   rw [Function.comp_apply, exact x |>.mpr H, h0]
 
-lemma Exact.of_comp_eq_zero_of_ker_in_range [Zero P] (hc : g.comp f = 0)
+lemma of_comp_eq_zero_of_ker_in_range [Zero P] (hc : g.comp f = 0)
     (hr : ∀ y, g y = 0 → y ∈ Set.range f) :
     Exact f g :=
   fun y ↦ ⟨hr y, fun ⟨x, hx⟩ ↦ hx ▸ congrFun hc x⟩
 
+end Exact
+
 end Function
 
+section AddMonoidHom
+
+variable [AddGroup M] [AddGroup N] [AddGroup P] {f : M →+ N} {g : N →+ P}
+
+namespace AddMonoidHom
+
+open Function
+
+lemma exact_iff :
+    Exact f g ↔ ker g = range f :=
+  Iff.symm SetLike.ext_iff
+
+lemma exact_of_comp_eq_zero_of_ker_le_range
+    (h1 : g.comp f = 0) (h2 : ker g ≤ range f) : Exact f g :=
+  Exact.of_comp_of_mem_range (congrArg DFunLike.coe h1) h2
+
+lemma exact_of_comp_of_mem_range
+    (h1 : g.comp f = 0) (h2 : ∀ x, g x = 0 → x ∈ range f) : Exact f g :=
+  exact_of_comp_eq_zero_of_ker_le_range h1 h2
+
+end AddMonoidHom
+
+namespace Function.Exact
+
+open AddMonoidHom
+
+lemma addMonoidHom_ker_eq (hfg : Exact f g) :
+    ker g = range f :=
+  SetLike.ext hfg
+
+lemma addMonoidHom_comp_eq_zero (h : Exact f g) : g.comp f = 0 :=
+  DFunLike.coe_injective h.comp_eq_zero
+
+section
+
+variable {X₁ X₂ X₃ Y₁ Y₂ Y₃ : Type*} [AddCommMonoid X₁] [AddCommMonoid X₂] [AddCommMonoid X₃]
+  [AddCommMonoid Y₁] [AddCommMonoid Y₂] [AddCommMonoid Y₃]
+  (e₁ : X₁ ≃+ Y₁) (e₂ : X₂ ≃+ Y₂) (e₃ : X₃ ≃+ Y₃)
+  {f₁₂ : X₁ →+ X₂} {f₂₃ : X₂ →+ X₃} {g₁₂ : Y₁ →+ Y₂} {g₂₃ : Y₂ →+ Y₃}
+  (comm₁₂ : g₁₂.comp e₁.toAddMonoidHom = e₂.toAddMonoidHom.comp f₁₂)
+  (comm₂₃ : g₂₃.comp e₂.toAddMonoidHom = e₃.toAddMonoidHom.comp f₂₃)
+
+lemma of_ladder_addEquiv_of_exact (H : Exact f₁₂ f₂₃) : Exact g₁₂ g₂₃ := by
```

Comment:
Should the name mention some `comm` stuff?

Reply:
I would think that this was the intent of `ladder` in the original name of the `LinearMap` version, which I have basically duplicated here for `AddMonoidHom`.